### PR TITLE
added check for circular dependencies

### PR DESCRIPTION
**Pull request by etbyrd**
**Customer scenario**

In certain circumstances, there can be a circular dependency that creates a stack overflow when recursively checking for unresolved child dependencies, causing a complete VS crash. 

**Bugs this fixes:** 

https://devdiv.visualstudio.com/DevDiv/_workitems/edit/487858
https://github.com/dotnet/project-system/issues/3374

**Workarounds, if any**

None

**Risk**

Low, there are some threading concerns that I am looking into to make sure that the Hashset is not shared - but I do not think that will be an issue because I am editing a private method and the class is instantiated when creating a snapshot.

**Performance impact**

Low. Just adding a HashSet. 

**Is this a regression from a previous update?**

No

**Root cause analysis:**

There was no check that a dependency was not already resolved. 

**How was the bug found?**

Customer feedback.

### CODE REVIEW

**Comment by jmarolf** [comment]
@etbyrd is there a way to write a test for this case?

**Comment by etbyrd** [comment]
retest windows_release please

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshot.cs [hunk 1]
```diff
@@ -55,6 +55,8 @@ internal class TargetedDependenciesSnapshot : ITargetedDependenciesSnapshot
         private Dictionary<string, bool> _unresolvedDescendantsMap
             = new Dictionary<string, bool>(StringComparer.OrdinalIgnoreCase);
 
+        private HashSet<string> _seenDependencies = new HashSet<string>();
```
Comment (abpiskunov) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshot.cs]: > private HashSet<string> _seenDependencies = new HashSet<string>(); [](start = 7, length = 67)

Why keep it global? It seems like implementation detail of FindUnresolvedDependenciesRecursive and don't have to stay in memory? We already keeping too much stuff in memory ...

Comment (etbyrd) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshot.cs]: Good point. I will fix that

**Comment by etbyrd** [comment]
@jmarolf @abpiskunov I will be adding tests for this - I am still worried about some edge cases

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshot.cs [hunk 2]
```diff
@@ -55,6 +55,8 @@ internal class TargetedDependenciesSnapshot : ITargetedDependenciesSnapshot
         private Dictionary<string, bool> _unresolvedDescendantsMap
             = new Dictionary<string, bool>(StringComparer.OrdinalIgnoreCase);
 
+        
+
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshot.cs]: Nit: remove this added blank space.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshotTests.cs [hunk 3]
```diff
@@ -573,6 +573,47 @@ public void TFromChanges_RemovedAndAddedChanges()
             Assert.True(snapshot.DependenciesWorld.ContainsKey(@"tfm1\xxx\addeddependency3"));
         }
 
+        /// <summary>
+        /// Added because circular dependencies can cause stack overflows
+        /// https://github.com/dotnet/project-system/pull/3474
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshotTests.cs]: Link to [https://github.com/dotnet/project-system/issues/3374](https://github.com/dotnet/project-system/issues/3374) instead please

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshot.cs [hunk 4]
```diff
@@ -32,6 +32,20 @@ internal class TargetedDependenciesSnapshot : ITargetedDependenciesSnapshot
             }
         }
 
+        /// <summary>
+        /// Internal for testing.
+        /// </summary>
+        internal TargetedDependenciesSnapshot(IDictionary<string, IDependency> dependenciesWorld, IEnumerable<IDependency> topLevelDependencies)
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshot.cs]: I would rather make this constructor private, with in `internal static` factory method with a clear name like `CreateForUnitTestingOnly` or something.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshot.cs [hunk 5]
```diff
@@ -113,14 +127,23 @@ public IEnumerable<IDependency> GetDependencyChildren(IDependency dependency)
...
-                        result = true;
+                        unresolved = true;
+                        break;
+                    }
+
+                    //If the dependency is already in the child map, it is resolved
+                    //Checking here will prevent a stack overflow due to rechecking the same dependencies
+                    if (_dependenciesChildrenMap.ContainsKey(child.Id))
+                    {
+                        unresolved = false;
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshot.cs]: I had to think about why this is sufficient to break cycles.  My conclusion is that we don't have to worry about the unresolved case, because unresolved dependencies wouldn't have any children, is that right?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshot.cs [hunk 6]
```diff
@@ -113,14 +127,23 @@ public IEnumerable<IDependency> GetDependencyChildren(IDependency dependency)
...
                 foreach (var child in GetDependencyChildren(dependency))
                 {
                     if (!child.Resolved)
                     {
-                        result = true;
+                        unresolved = true;
+                        break;
+                    }
+
+                    //If the dependency is already in the child map, it is resolved
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshot.cs]: Nit: We put a space after `//` in comments.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshotTests.cs [hunk 7]
```diff
@@ -573,6 +573,47 @@ public void TFromChanges_RemovedAndAddedChanges()
...
+            var previousSnapshot = new TargetedDependenciesSnapshot(
+                dependenciesWorld: new Dictionary<string, IDependency>()
+                {
+                    { dependencyModelTop1.Id, dependencyModelTop1 },
+                    { dependencyModelTop2.Id, dependencyModelTop2 },
+                },
+                topLevelDependencies: new List<IDependency>() { dependencyModelTop1 });
+
+            // verify it doesn't stack overflow
+            previousSnapshot.CheckForUnresolvedDependencies(dependencyModelTop1);
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshotTests.cs]: Have you verified that this test fails without your fix?

Comment (etbyrd) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshotTests.cs]: Yes, it fails without the fix.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshotTests.cs (hunk 3 above)
Comment (etbyrd) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshotTests.cs]: Done

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshotTests.cs [hunk 8]
```diff
@@ -573,6 +573,47 @@ public void TFromChanges_RemovedAndAddedChanges()
...
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshotTests.cs]: I would prefer if we didn't have the failure case here be: crash the test harness.  That is going to be hard to diagnose if we ever regress this.

Comment (etbyrd) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshotTests.cs]: I agree. I am working a way around this but I wonder if it is a requirement for this fix to be added.
